File: frontend/src/types/domain.ts
export interface Coordinates {
  latitude: number;
  longitude: number;
}

export interface City {
  id: string;
  name: string;
  state: string;
  country: string;
  timezone: string;
}

export interface Ward {
  id: string;
  city_id: string;
  name: string;
  code?: string;
  centroid?: Coordinates;
  population?: number;
}

export interface MonitoringStation {
  id: string;
  city_id: string;
  ward_id?: string;
  name: string;
  location: Coordinates;
  operator?: string;
  station_type?: string;
  is_active: boolean;
}

export interface AirQualityObservation {
  station_id: string;
  observed_at: string;
  aqi?: number;
  pm25_ug_m3?: number;
  pm10_ug_m3?: number;
  no2_ug_m3?: number;
  so2_ug_m3?: number;
  co_mg_m3?: number;
  o3_ug_m3?: number;
  source: string;
  quality_flag?: string;
}

export interface WeatherObservation {
  location_id: string;
  observed_at: string;
  temperature_c?: number;
  relative_humidity_pct?: number;
  wind_speed_m_s?: number;
  wind_direction_deg?: number;
  precipitation_mm?: number;
  pressure_hpa?: number;
  boundary_layer_height_m?: number;
  source: string;
}

export interface AQIForecast {
  location_id: string;
  issued_at: string;
  valid_at: string;
  horizon_hours: number;
  predicted_aqi: number;
  lower_bound?: number;
  upper_bound?: number;
  model_version: string;
}

export type PollutionSourceCategory =
  | "traffic"
  | "construction"
  | "industrial"
  | "waste_burning"
  | "residential"
  | "dust"
  | "regional_transport"
  | "other";

export interface PollutionSourceAttribution {
  location_id: string;
  estimated_at: string;
  pollutant: string;
  contributions: Array<{
    category: PollutionSourceCategory;
    contribution_pct: number;
    confidence?: number;
  }>;
  method: string;
  model_version: string;
}

export interface InterventionRecommendation {
  id: string;
  location_id: string;
  created_at: string;
  title: string;
  rationale: string;
  recommended_actions: string[];
  priority: "low" | "medium" | "high" | "critical";
  responsible_agencies: string[];
  expected_impact?: string;
  evidence_refs: string[];
}

export interface CitizenAdvisory {
  id: string;
  location_id: string;
  issued_at: string;
  valid_until: string;
  language: string;
  risk_level: string;
  headline: string;
  guidance: string[];
  audience_groups: string[];
}

export type VulnerableLocationType =
  | "hospital"
  | "school"
  | "elder_care"
  | "child_care"
  | "outdoor_work_zone"
  | "other";

export interface VulnerableLocation {
  id: string;
  city_id: string;
  ward_id?: string;
  name: string;
  location: Coordinates;
  location_type: VulnerableLocationType;
  estimated_people_exposed?: number;
  data_source: string;
}

export interface HealthResponse {
  status: "ok";
  service: string;
  version: string;
}

export interface ProjectInfo {
  name: string;
  problem_statement: string;
  implementation_stage: string;
  planned_modules: string[];
}

File: backend/app/schemas/domain.py
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field


class AirViewSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")


class Coordinates(AirViewSchema):
    latitude: float = Field(ge=-90, le=90)
    longitude: float = Field(ge=-180, le=180)


class City(AirViewSchema):
    id: str
    name: str
    state: str
    country: str = "India"
    timezone: str = "Asia/Kolkata"


class Ward(AirViewSchema):
    id: str
    city_id: str
    name: str
    code: str | None = None
    centroid: Coordinates | None = None
    population: int | None = Field(default=None, ge=0)


class MonitoringStation(AirViewSchema):
    id: str
    city_id: str
    ward_id: str | None = None
    name: str
    location: Coordinates
    operator: str | None = None
    station_type: str | None = None
    is_active: bool = True


class AirQualityObservation(AirViewSchema):
    station_id: str
    observed_at: datetime
    aqi: float | None = Field(default=None, ge=0)
    pm25_ug_m3: float | None = Field(default=None, ge=0)
    pm10_ug_m3: float | None = Field(default=None, ge=0)
    no2_ug_m3: float | None = Field(default=None, ge=0)
    so2_ug_m3: float | None = Field(default=None, ge=0)
    co_mg_m3: float | None = Field(default=None, ge=0)
    o3_ug_m3: float | None = Field(default=None, ge=0)
    source: str
    quality_flag: str | None = None


class WeatherObservation(AirViewSchema):
    location_id: str
    observed_at: datetime
    temperature_c: float | None = None
    relative_humidity_pct: float | None = Field(default=None, ge=0, le=100)
    wind_speed_m_s: float | None = Field(default=None, ge=0)
    wind_direction_deg: float | None = Field(default=None, ge=0, lt=360)
    precipitation_mm: float | None = Field(default=None, ge=0)
    pressure_hpa: float | None = Field(default=None, ge=0)
    boundary_layer_height_m: float | None = Field(default=None, ge=0)
    source: str


class AQIForecast(AirViewSchema):
    location_id: str
    issued_at: datetime
    valid_at: datetime
    horizon_hours: int = Field(ge=1, le=72)
    predicted_aqi: float = Field(ge=0)
    lower_bound: float | None = Field(default=None, ge=0)
    upper_bound: float | None = Field(default=None, ge=0)
    model_version: str


class SourceCategory(str, Enum):
    TRAFFIC = "traffic"
    CONSTRUCTION = "construction"
    INDUSTRIAL = "industrial"
    WASTE_BURNING = "waste_burning"
    RESIDENTIAL = "residential"
    DUST = "dust"
    REGIONAL_TRANSPORT = "regional_transport"
    OTHER = "other"


class SourceContribution(AirViewSchema):
    category: SourceCategory
    contribution_pct: float = Field(ge=0, le=100)
    confidence: float | None = Field(default=None, ge=0, le=1)


class PollutionSourceAttribution(AirViewSchema):
    location_id: str
    estimated_at: datetime
    pollutant: str
    contributions: list[SourceContribution]
    method: str
    model_version: str


class InterventionPriority(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class InterventionRecommendation(AirViewSchema):
    id: str
    location_id: str
    created_at: datetime
    title: str
    rationale: str
    recommended_actions: list[str]
    priority: InterventionPriority
    responsible_agencies: list[str] = Field(default_factory=list)
    expected_impact: str | None = None
    evidence_refs: list[str] = Field(default_factory=list)


class CitizenAdvisory(AirViewSchema):
    id: str
    location_id: str
    issued_at: datetime
    valid_until: datetime
    language: str
    risk_level: str
    headline: str
    guidance: list[str]
    audience_groups: list[str] = Field(default_factory=list)


class VulnerableLocationType(str, Enum):
    HOSPITAL = "hospital"
    SCHOOL = "school"
    ELDER_CARE = "elder_care"
    CHILD_CARE = "child_care"
    OUTDOOR_WORK_ZONE = "outdoor_work_zone"
    OTHER = "other"


class VulnerableLocation(AirViewSchema):
    id: str
    city_id: str
    ward_id: str | None = None
    name: str
    location: Coordinates
    location_type: VulnerableLocationType
    estimated_people_exposed: int | None = Field(default=None, ge=0)
    data_source: str
